src/Compilers/CSharp/Portable/BoundTree/BoundNodeExtensions.cs

```diff
@@ -94,5 +97,163 @@ private class ContainsAwaitVisitor : BoundTreeWalkerWithStackGuardWithoutRecursi
+                {
+                    case BoundBinaryOperator op:
+                        binaryOperatorCallback?.Invoke(op, arg);
+                        break;
+                    case TInterpolatedStringType interpolatedString:
+                        if (!visitor(interpolatedString, arg))
+                        {
+                            return false;
+                        }
+                        break;
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(current.Left.Kind);
+                }
+
+                switch (current.Right)
+                {
+                    case BoundBinaryOperator rightOperator:
+                        binaryOperatorCallback?.Invoke(rightOperator, arg);
+                        pushLeftNodes(rightOperator, stack);
+                        break;
+                    case TInterpolatedStringType interpolatedString:
+                        if (!visitor(interpolatedString, arg))
+                        {
+                            return false;
+                        }
+                        break;
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(current.Right.Kind);
+                }
+            }
+
+            Debug.Assert(stack.Count == 0);
+            stack.Free();
+            return true;
+
+            static void pushLeftNodes(BoundBinaryOperator binary, ArrayBuilder<BoundBinaryOperator> stack)
+            {
+                Debug.Assert(typeof(TInterpolatedStringType) == typeof(BoundInterpolatedString) || binary.IsUnconvertedInterpolatedStringAddition);
+                BoundBinaryOperator? current = binary;
+                while (current != null)
+                {
+                    stack.Push(current);
+                    current = current.Left as BoundBinaryOperator;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Rewrites a BoundBinaryOperator composed of interpolated strings (either converted or unconverted) iteratively, without
+        /// recursion.
+        /// </summary>
+        /// <param name="binary">The original top of the binary operations.</param>
+        /// <param name="arg">The callback args.</param>
+        /// <param name="interpolatedStringFactory">
+        /// Rewriter for the BoundInterpolatedString or BoundUnconvertedInterpolatedString parts of the binary operator. Passed the callback
+        /// parameter, the original interpolated string, and the index of the interpolated string in the tree.
+        /// </param>
+        /// <param name="binaryOperatorFactory">
+        /// Rewriter for the BoundBinaryOperator parts fo the binary operator. Passed the callback parameter, the original binary operator, and
+        /// the rewritten left and right components.
+        /// </param>
+        public static TResult RewriteInterpolatedStringAddition<TArg, TInterpolatedStringType, TResult>(
```

Comment:
Would it be simpler if we leveraged `VisitBinaryOperatorInterpolatedString` and avoided a separate implementation of traversal? 
If I understood the traversal order of `VisitBinaryOperatorInterpolatedString` correctly, then the following should work: 
  When we encounter a leaf, we rewrite it and push the rewritten node to a stack. 
  When we encounter a binary node, we pop two rewritten nodes from the stack, rewrite the binary node with those, and push the rewritten binary node.
  After the traversal is complete, the stack should contain exactly one entry for the rewrittern result.

Having a visit helper with a single callback for all nodes (as suggested above) might make that logic even clearer in terms of ordering.

(I didn't review fully this method yet)

Reply:
We talked offline. The Visit version needs to be depth-first in-order, so it can't be used here. Instead, I'm going to switch this to only use iteration for the left side of the tree, and recurse on the right side when necessary. Since we are generally only concerned with large trees on the left side, this should be no worse than other sections of our code for stack behavior.